File: backend/app/models/enums.py
"""Core enumeration types for CapVeri domain models.

These enums use the str mixin to ensure JSON serialization works correctly.
Values must match exactly with frontend/src/types/enums.ts.
"""

from enum import Enum


class AccountingBasis(str, Enum):
    """Accounting basis for GL entry date filtering.

    Cash: only expenses physically paid during the period are recoverable.
    Accrual: expenses incurred during the period count regardless of payment date.
    """

    CASH = "cash"
    ACCRUAL = "accrual"


class CapType(str, Enum):
    """Type of expense cap applied to tenant recoveries."""

    NONE = "none"
    NON_CUMULATIVE = "non_cumulative"
    CUMULATIVE = "cumulative"
    CUMULATIVE_COMPOUNDING = "cumulative_compounding"


class PoolType(str, Enum):
    """Category of expense pool for allocation."""

    OPERATING = "operating"
    TAX = "tax"
    INSURANCE = "insurance"
    CAPITAL = "capital"
    OTHER = "other"


class AllocationType(str, Enum):
    """Type of allocation for splitting expense pools."""

    PERCENTAGE = "percentage"
    FIXED_AMOUNT = "fixed_amount"


class LeaseStatus(str, Enum):
    """Current status of a lease."""

    DRAFT = "draft"
    ACTIVE = "active"
    EXPIRED = "expired"
    TERMINATED = "terminated"


class ImportStatus(str, Enum):
    """Status of a data import batch."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class UserRole(str, Enum):
    """Role of a user within an organization."""

    OWNER = "owner"
    ADMIN = "admin"
    MEMBER = "member"
    VIEWER = "viewer"
    TENANT = "tenant"


class ReconciliationStatus(str, Enum):
    """Status of a reconciliation snapshot."""

    DRAFT = "draft"
    FINALIZED = "finalized"


class CampaignStatus(str, Enum):
    """Workflow status of a reconciliation campaign (property-year lifecycle)."""

    DRAFT = "draft"
    FINALIZED = "finalized"
    IN_REVIEW = "in_review"
    APPROVED = "approved"
    SENT = "sent"


class UnitStatus(str, Enum):
    """Status of a unit within a property."""

    VACANT = "vacant"
    OCCUPIED = "occupied"
    UNDER_RENOVATION = "under_renovation"


class DocumentStatus(str, Enum):
    """Status of a document in the OCR pipeline."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    READY_FOR_REVIEW = "ready_for_review"
    VERIFIED = "verified"
    REJECTED = "rejected"


class DocumentType(str, Enum):
    """Type of document for OCR processing."""

    LEASE = "lease"
    AMENDMENT = "amendment"
    RENT_ROLL = "rent_roll"
    GL_EXPORT = "gl_export"
    OTHER = "other"


class ExtractionJobStatus(str, Enum):
    """Status of an extraction job in the async queue."""

    PENDING = "pending"  # Job queued, not yet started
    PROCESSING = "processing"  # Job currently being processed
    COMPLETED = "completed"  # Job completed successfully
    FAILED = "failed"  # Job failed after retries
    RETRYING = "retrying"  # Job failed, will retry


class ExtractionJobPriority(int, Enum):
    """Priority level for extraction jobs (higher = more urgent)."""

    LOW = 0
    NORMAL = 5
    HIGH = 10
    URGENT = 15


class StatementStatus(str, Enum):
    """Status of a tenant reconciliation statement."""

    PENDING = "pending"
    PAID = "paid"
    DISPUTED = "disputed"
    OVERDUE = "overdue"


class NotificationType(str, Enum):
    """Type of tenant notification."""

    NEW_STATEMENT = "new_statement"
    DISPUTE_UPDATE = "dispute_update"
    STATEMENT_REMINDER = "statement_reminder"
    SYSTEM = "system"


class DisputeStatus(str, Enum):
    """Status of a dispute in the workflow."""

    OPEN = "open"
    UNDER_REVIEW = "under_review"
    RESOLVED = "resolved"
    REJECTED = "rejected"
    CLOSED = "closed"


class DisputeCategory(str, Enum):
    """Category of dispute issue."""

    CALCULATION_ERROR = "calculation_error"
    MISSING_CREDIT = "missing_credit"
    INCORRECT_AREA = "incorrect_area"
    BASE_YEAR_ISSUE = "base_year_issue"
    BILLING_QUESTION = "billing_question"
    OTHER = "other"


class CalculationJobStatus(str, Enum):
    """Status of a calculation job."""

    PENDING = "pending"  # Job created, waiting to start
    RUNNING = "running"  # Job currently executing
    COMPLETED = "completed"  # Job finished successfully
    FAILED = "failed"  # Job failed with error


class ERPFormat(str, Enum):
    """Format for ERP write-back export."""

    YARDI = "yardi"  # Yardi Voyager format
    MRI = "mri"  # MRI Commercial format
    CSV = "csv"  # Generic CSV format


class BomaStandardVersion(str, Enum):
    """BOMA Office Standard version used for area measurements."""

    V2010 = "2010"
    V2017 = "2017"
    V2024 = "2024"
    CUSTOM = "custom"


class SpaceType(str, Enum):
    """BOMA 2024 space classification for units.

    NATA (Non-Allocated Tenant Areas) types — storage, outdoor_amenity,
    equipment_shaft — must have zero load factor in pro-rata calculations
    per BOMA 2024 Office Standard.
    """

    OFFICE = "office"
    RETAIL = "retail"
    LABORATORY = "laboratory"
    STORAGE = "storage"  # NATA — zero load factor
    OUTDOOR_AMENITY = "outdoor_amenity"  # NATA — zero load factor
    EQUIPMENT_SHAFT = "equipment_shaft"  # NATA — zero load factor
    OTHER = "other"


# NATA space types require zero load factor in pro-rata calculations (BOMA 2024)
NATA_SPACE_TYPES: frozenset[SpaceType] = frozenset(
    {
        SpaceType.STORAGE,
        SpaceType.OUTDOOR_AMENITY,
        SpaceType.EQUIPMENT_SHAFT,
    }
)

File: frontend/src/types/enums.ts
/**
 * Core enumeration types for CapVeri domain models.
 *
 * These const objects use the "as const" pattern for type-safe enums.
 * Values must match exactly with backend/app/models/enums.py.
 */

/**
 * Type of expense cap applied to tenant recoveries.
 */
export const CapType = {
  NONE: 'none',
  NON_CUMULATIVE: 'non_cumulative',
  CUMULATIVE: 'cumulative',
  CUMULATIVE_COMPOUNDING: 'cumulative_compounding',
} as const
export type CapType = (typeof CapType)[keyof typeof CapType]

/**
 * Category of expense pool for allocation.
 */
export const PoolType = {
  OPERATING: 'operating',
  TAX: 'tax',
  INSURANCE: 'insurance',
  CAPITAL: 'capital',
  OTHER: 'other',
} as const
export type PoolType = (typeof PoolType)[keyof typeof PoolType]

/**
 * Type of allocation for splitting expense pools.
 */
export const AllocationType = {
  PERCENTAGE: 'percentage',
  FIXED_AMOUNT: 'fixed_amount',
} as const
export type AllocationType =
  (typeof AllocationType)[keyof typeof AllocationType]

/**
 * Accounting basis for GL entry date filtering per lease.
 */
export const AccountingBasis = {
  CASH: 'cash',
  ACCRUAL: 'accrual',
} as const
export type AccountingBasis =
  (typeof AccountingBasis)[keyof typeof AccountingBasis]

/**
 * Current status of a lease.
 */
export const LeaseStatus = {
  DRAFT: 'draft',
  ACTIVE: 'active',
  EXPIRED: 'expired',
  TERMINATED: 'terminated',
} as const
export type LeaseStatus = (typeof LeaseStatus)[keyof typeof LeaseStatus]

/**
 * Status of a data import batch.
 */
export const ImportStatus = {
  PENDING: 'pending',
  PROCESSING: 'processing',
  COMPLETED: 'completed',
  FAILED: 'failed',
} as const
export type ImportStatus = (typeof ImportStatus)[keyof typeof ImportStatus]

/**
 * Role of a user within an organization.
 */
export const UserRole = {
  OWNER: 'owner',
  ADMIN: 'admin',
  MEMBER: 'member',
  VIEWER: 'viewer',
  TENANT: 'tenant',
} as const
export type UserRole = (typeof UserRole)[keyof typeof UserRole]

/**
 * Status of a reconciliation snapshot.
 */
export const ReconciliationStatus = {
  DRAFT: 'draft',
  FINALIZED: 'finalized',
} as const
export type ReconciliationStatus =
  (typeof ReconciliationStatus)[keyof typeof ReconciliationStatus]

/**
 * Workflow status of a reconciliation campaign (property-year lifecycle).
 */
export const CampaignStatus = {
  DRAFT: 'draft',
  FINALIZED: 'finalized',
  IN_REVIEW: 'in_review',
  APPROVED: 'approved',
  SENT: 'sent',
} as const
export type CampaignStatus =
  (typeof CampaignStatus)[keyof typeof CampaignStatus]

/**
 * Status of a unit within a property.
 */
export const UnitStatus = {
  VACANT: 'vacant',
  OCCUPIED: 'occupied',
  UNDER_RENOVATION: 'under_renovation',
} as const
export type UnitStatus = (typeof UnitStatus)[keyof typeof UnitStatus]

/**
 * Status of a document in the OCR pipeline.
 */
export const DocumentStatus = {
  PENDING: 'pending',
  PROCESSING: 'processing',
  COMPLETED: 'completed',
  FAILED: 'failed',
  READY_FOR_REVIEW: 'ready_for_review',
  VERIFIED: 'verified',
  REJECTED: 'rejected',
} as const
export type DocumentStatus =
  (typeof DocumentStatus)[keyof typeof DocumentStatus]

/**
 * Status of a tenant reconciliation statement.
 */
export const StatementStatus = {
  PENDING: 'pending',
  PAID: 'paid',
  DISPUTED: 'disputed',
  OVERDUE: 'overdue',
} as const
export type StatementStatus =
  (typeof StatementStatus)[keyof typeof StatementStatus]

/**
 * Type of document for OCR processing.
 */
export const DocumentType = {
  LEASE: 'lease',
  AMENDMENT: 'amendment',
  RENT_ROLL: 'rent_roll',
  GL_EXPORT: 'gl_export',
  OTHER: 'other',
} as const
export type DocumentType = (typeof DocumentType)[keyof typeof DocumentType]

/**
 * Status of an extraction job in the async queue.
 */
export const ExtractionJobStatus = {
  PENDING: 'pending',
  PROCESSING: 'processing',
  COMPLETED: 'completed',
  FAILED: 'failed',
  RETRYING: 'retrying',
} as const
export type ExtractionJobStatus =
  (typeof ExtractionJobStatus)[keyof typeof ExtractionJobStatus]

/**
 * Priority level for extraction jobs (higher = more urgent).
 */
export const ExtractionJobPriority = {
  LOW: 0,
  NORMAL: 5,
  HIGH: 10,
  URGENT: 15,
} as const
export type ExtractionJobPriority =
  (typeof ExtractionJobPriority)[keyof typeof ExtractionJobPriority]

/**
 * Type of tenant notification.
 */
export const NotificationType = {
  NEW_STATEMENT: 'new_statement',
  DISPUTE_UPDATE: 'dispute_update',
  STATEMENT_REMINDER: 'statement_reminder',
  SYSTEM: 'system',
} as const
export type NotificationType =
  (typeof NotificationType)[keyof typeof NotificationType]

/**
 * Status of a dispute in the workflow.
 */
export const DisputeStatus = {
  OPEN: 'open',
  UNDER_REVIEW: 'under_review',
  RESOLVED: 'resolved',
  REJECTED: 'rejected',
  CLOSED: 'closed',
} as const
export type DisputeStatus = (typeof DisputeStatus)[keyof typeof DisputeStatus]

/**
 * Category of dispute issue.
 */
export const DisputeCategory = {
  CALCULATION_ERROR: 'calculation_error',
  MISSING_CREDIT: 'missing_credit',
  INCORRECT_AREA: 'incorrect_area',
  BASE_YEAR_ISSUE: 'base_year_issue',
  BILLING_QUESTION: 'billing_question',
  OTHER: 'other',
} as const
export type DisputeCategory =
  (typeof DisputeCategory)[keyof typeof DisputeCategory]

/**
 * Status of a calculation job.
 */
export const CalculationJobStatus = {
  PENDING: 'pending',
  RUNNING: 'running',
  COMPLETED: 'completed',
  FAILED: 'failed',
} as const
export type CalculationJobStatus =
  (typeof CalculationJobStatus)[keyof typeof CalculationJobStatus]

/**
 * Status of an invoice in the billing system.
 */
export const InvoiceStatus = {
  DRAFT: 'draft',
  OPEN: 'open',
  PAID: 'paid',
  VOID: 'void',
  UNCOLLECTIBLE: 'uncollectible',
} as const
export type InvoiceStatus = (typeof InvoiceStatus)[keyof typeof InvoiceStatus]

/**
 * Type of user feedback.
 */
export const FeedbackType = {
  BUG: 'bug',
  FEATURE_REQUEST: 'feature_request',
  GENERAL: 'general',
} as const
export type FeedbackType = (typeof FeedbackType)[keyof typeof FeedbackType]

/**
 * Status of user feedback.
 */
export const FeedbackStatus = {
  NEW: 'new',
  REVIEWED: 'reviewed',
  RESOLVED: 'resolved',
  DISMISSED: 'dismissed',
} as const
export type FeedbackStatus =
  (typeof FeedbackStatus)[keyof typeof FeedbackStatus]

/**
 * Organization subscription status.
 */
export const SubscriptionStatus = {
  ACTIVE: 'active',
  TRIAL: 'trial',
  SUSPENDED: 'suspended',
  CANCELLED: 'cancelled',
} as const
export type SubscriptionStatus =
  (typeof SubscriptionStatus)[keyof typeof SubscriptionStatus]

/**
 * Stripe billing subscription status.
 */
export const BillingSubscriptionStatus = {
  TRIALING: 'trialing',
  ACTIVE: 'active',
  PAST_DUE: 'past_due',
  CANCELED: 'canceled',
  PAUSED: 'paused',
} as const
export type BillingSubscriptionStatus =
  (typeof BillingSubscriptionStatus)[keyof typeof BillingSubscriptionStatus]

/**
 * Type of promotional discount.
 */
export const DiscountType = {
  PERCENTAGE: 'percentage',
  FIXED_AMOUNT: 'fixed_amount',
  FREE_TRIAL_EXTENSION: 'free_trial_extension',
} as const
export type DiscountType = (typeof DiscountType)[keyof typeof DiscountType]

/**
 * Status of a promotional offer.
 */
export const PromotionStatus = {
  ACTIVE: 'active',
  EXPIRED: 'expired',
  EXHAUSTED: 'exhausted',
  DISABLED: 'disabled',
} as const
export type PromotionStatus =
  (typeof PromotionStatus)[keyof typeof PromotionStatus]

/**
 * Format for ERP export files.
 */
export const ERPFormat = {
  YARDI: 'yardi',
  MRI: 'mri',
  CSV: 'csv',
} as const
export type ERPFormat = (typeof ERPFormat)[keyof typeof ERPFormat]
